### PR TITLE
add a FAQ

### PR DESCRIPTION
**Pull request by basnijholt**
@jbweston and @akhmerov, feel free to edit and add answers and questions.

### CODE REVIEW

File: docs/source/faq.rst [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+Use the `adaptive.DataSaver`.
+
+
+My runner failed, how do I get the error message?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Check ``runner.task.print_stack()``.
+
+
+How do I get a `~adaptiveLearner2D`\'s data on a grid?
```
Comment (jbweston) [inline docs/source/faq.rst]:
```suggestion
How do I get a `~adaptive.Learner2D`\'s data on a grid?
```

File: docs/source/faq.rst [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use ``learner.interpolated_on_grid()`` optionally with a argument ``n`` to specify the the amount of points in ``x`` and ``y``.
+
+
+Why can I not use a ``lambda`` function?
```
Comment (jbweston) [inline docs/source/faq.rst]:
```suggestion
Why can I not use a ``lambda`` with a learner?
```

File: docs/source/faq.rst [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The functions evaluations are done on an executor.
```
Comment (jbweston) [inline docs/source/faq.rst]:
```suggestion
When using the `~adaptive.Runner` the learner's function is evaluated in different Python processes.
```

File: docs/source/faq.rst [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+Therefore, the ``function`` needs to be serialized (pickled) and send to the executor, ``lambda``\s cannot be pickled.
```
Comment (jbweston) [inline docs/source/faq.rst]:
```suggestion
Therefore, the ``function`` needs to be serialized (pickled) and send to the other Python processes; ``lambda``\s cannot be pickled.
```

File: docs/source/faq.rst [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Read our `draft paper <https://gitlab.kwant-project.org/qt/adaptive-paper/builds/artifacts/master/file/paper.pdf?job=make>`_ or the source code on `GitHub <https://github.com/python-adaptive/adaptive/>`_.
+
+
+How do I run multiple runners?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Check out `Adaptive scheduler <http://adaptive-scheduler.readthedocs.io>`_, which solves the following problem of needing to run more learners than you can run with a single runner.
+It easily runs on tens of thousands of cores.
```
Comment (jbweston) [inline docs/source/faq.rst]: I think this question/answer needs more context.

You can absolutely run more than 1 runner at a time in a notebook in the obvious way:

```python
runner1 = Runner(learner1)
runner2 = Runner(learner2)
```

File: docs/source/faq.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+
+What is the difference with FEM?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The main difference with FEM is that one needs to globally update the mesh at every time step.
```
Comment (jbweston) [inline docs/source/faq.rst]: Need to define what FEM means.

The only acronym I know is "Finite Element Method", which is something totally unrelated

Comment (jbweston) [reply docs/source/faq.rst]: And the explanation below would need to be updated to reflect this.

File: docs/source/faq.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+What is the difference with adaptive meshing in CFD or computer graphics?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+XXX: add answer!
+
+
+Can I use this to tune my hyper parameters for machine learning models?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+XXX: add answer!
```
Comment (jbweston) [inline docs/source/faq.rst]: Was this question ever asked by a user? I would classify this as an "open research question"

File: example-notebook.ipynb [hunk 8]
```diff
@@ -104,7 +104,7 @@
    "source": [
     "# The end condition is when the \"loss\" is less than 0.01. In the context of the\n",
     "# 1D learner this means that we will resolve features in 'func' with width 0.01 or wider.\n",
-    "runner = adaptive.Runner(learner, goal=lambda l: l.loss() < 0.01, ntasks=4)\n",
+    "runner = adaptive.Runner(learner, goal=lambda l: l.loss() < 0.01)\n",
```
Comment (jbweston) [inline example-notebook.ipynb]:
```suggestion
    "runner = adaptive.Runner(learner, goal=lambda l: l.loss() < 0.01, tasks=4)\n",
```

This is unrelated to the FAQ

File: docs/source/faq.rst (hunk 7 above)
Comment (basnijholt) [reply docs/source/faq.rst]: This was asked a few dozen times on Reddit and also by at least one actual user.

File: docs/source/faq.rst (hunk 6 above)
Comment (basnijholt) [reply docs/source/faq.rst]: I am referring to "Finite Element Method" here.

File: docs/source/faq.rst (hunk 5 above)
Comment (basnijholt) [reply docs/source/faq.rst]: The problem I want to address is when the learner overhead is too large, such that one needs multiple runners in different threads/processes.

For example, when you use a `BalancingLearner` with 10.000 learners.

**Comment by jbweston** [comment]
@basnijholt what's the status of this?

**Comment by basnijholt** [comment]
Waiting on review and some answers. We could merge without the unanswered questions though.

**Comment by jbweston** [comment]
> Waiting on review and some answers. We could merge without the unanswered questions though.

OK. I'll put this on my review backlog

**Comment by basnijholt** [comment]
I've enabled auto-merge on this PR and commented out the unanswered questions.
